enhancements/update/cvo-log-level-api.md

```diff
@@ -0,0 +1,327 @@
+[`OperatorStatus`][OperatorStatus] fields. The `ObservedGeneration` field is 
+used to give feedback regarding the last generation change the CVO has dealt
+with. There is currently no need to support additional fields.
+
+The `ClusterVersionOperator` CRD will be behind a new 
+`ClusterVersionOperatorConfiguration` FeatureGate that will be used to 
+control the development of the CVO configuration.
+
+[github.com/openshift/api/operator/v1]: https://github.com/openshift/api/tree/master/operator/v1
+[OperatorSpec]: https://github.com/openshift/api/blob/f89ab92f1597eaed4de5b947c1781adde2bf42fb/operator/v1/types.go#L54
+[OperatorStatus]: https://github.com/openshift/api/blob/f89ab92f1597eaed4de5b947c1781adde2bf42fb/operator/v1/types.go#L112
+
+The proposed contents of the new `operator/v1/types_clusterversion.go` file:
+
+```go
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterVersionOperator holds cluster-wide information about the Cluster Version Operator.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+// +openshift:file-pattern=cvoRunLevel=0000_00,operatorName=cluster-version-operator,operatorOrdering=01
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=clusterversionoperators,scope=Cluster
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/2044
+// +openshift:enable:FeatureGate=ClusterVersionOperatorConfiguration
+type ClusterVersionOperator struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata"`
+
+	// spec is the specification of the desired behavior of the Cluster Version Operator.
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ClusterVersionOperatorSpec `json:"spec"`
+
+	// status is the most recently observed status of the Cluster Version Operator.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status ClusterVersionOperatorStatus `json:"status"`
+}
+
+// ClusterVersionOperatorSpec is the specification of the desired behavior of the Cluster Version Operator.
+type ClusterVersionOperatorSpec struct {
+	// operatorLogLevel is an intent based logging for the operator itself.  It does not give fine grained control, but it is a
+	// simple way to manage coarse grained logging choices that operators have to interpret for themselves.
+	//
+	// Valid values are: "Normal", "Debug", "Trace", "TraceAll".
+	// Defaults to "Normal".
+	// +optional
+	// +kubebuilder:default=Normal
+	OperatorLogLevel LogLevel `json:"operatorLogLevel,omitempty"`
```

Comment:
When I change this as a user, what happens? What observes this, and what changes? Normally, you'd expect the log level of a project to be exposed via a flag, but I'm guessing that can't happen for CVO, so, how does it work inside the operator?

Reply:
What ideally happens is:
1. A user patches the `ClusterVersionOperator` `cluster` resource and sets the `OperatorLogLevel` from the level `Normal` to the level `Trace`, for example.
2. The CVO notices that the  `ClusterVersionOperator` `cluster` resource was modified. It updates the `observedGeneration` field.
3. The CVO starts printing its logs up to the Trace level inclusive.

---

> Normally, you'd expect the log level of a project to be exposed via a flag, but I'm guessing that can't happen for CVO, so, how does it work inside the operator?

I am not sure that I fully understand the question. The `OperatorLogLevel` field is for modifying the runtime logs of the CVO executable. You can currently specify the log level of the CVO executable using the `--v` flag:
```
$ ./_output/linux/amd64/cluster-version-operator --help | grep "log level"
  -v, --v Level                          number for the log level verbosity
```

However, a user currently can't dynamically modify the log level of a CVO in a live cluster, as the CVO overwrites any changes to its deployment. 

As for how the CVO will be able to change its log level dynamically, the CVO uses the [`klog`](https://github.com/kubernetes/klog) library to log data. The exact level can be changed dynamically when using the `klog` library. For example, there is an even [`SetLogLevel`](https://github.com/openshift/library-go/blob/master/pkg/operator/loglevel/util.go#L63-L103) function to set the log level dynamically defined in the `openshift/library-go` repository.